### PR TITLE
Add tracing for GRPC API calls (#821)

### PR DESCRIPTION
**Pull request by arunkumarchacko**
Cherrypick https://github.com/GoogleCloudDataproc/hadoop-connectors/pull/821 to 2.2.x branch

### CODE REVIEW

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GrpcRequestTracingInfo.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.hadoop.gcsio;
+
+import java.util.OptionalLong;
+
+public class GrpcRequestTracingInfo {
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GrpcRequestTracingInfo.java:19]: Make this class package-private and add javadoc

**Review by medb — CHANGES_REQUESTED** [review]

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcTracingInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import java.net.SocketAddress;
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+
+/** Interceptor to create a trace of the lifecycle of GRPC api calls. */
+public class GoogleCloudStorageGrpcTracingInterceptor implements io.grpc.ClientInterceptor {
```
[the remark this answers is not in the document]
Comment (arunkumarchacko) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcTracingInterceptor.java:37]: I agree with you that we should make this package-private. I had to make this public to enable testing (Integration tests use this to listen to the logs). One option is to add @VisibleForTesting annotation. WDYT?

[the remark this answers is not in the document]
Comment (medb) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcTracingInterceptor.java:37]: Yes, let's add `@VisibleForTesting`

**Review by medb — APPROVED** [review]

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcTracingInterceptor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (arunkumarchacko) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcTracingInterceptor.java:37]: Opened b/239642937 to track this. Will raise a separate PR in order to keep the cherrypick clean.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GrpcRequestTracingInfo.java (hunk 1 above)
Comment (arunkumarchacko) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GrpcRequestTracingInfo.java:19]: Opened b/239642937 to track this.

**Pull request merged**